### PR TITLE
Support for a bitcoind 'ready' file to indicate startup is complete.

### PR DESCRIPTION
**Pull request by murrayn**
Please see the discussion of #15367 for detailed rationale. Essentially, the idea is to indicate, via creation of a file, that `bitcoind` is not only running but ready to actually use. This should provide a convenient way for scripts to reliably start `bitcoind`, wait for it to become ready, and then subsequently start any other services (e.g. `lnd`) that depend on it.

This PR also adds a `-ready` argument for specifying the name of the ready file (default: `$DATADIR/.ready`)

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -194,6 +230,7 @@ void Interrupt()
 void Shutdown(InitInterfaces& interfaces)
 {
     LogPrintf("%s: In progress...\n", __func__);
+    RemoveFile(GetReadyFile());
```
Comment (luke-jr) [inline src/init.cpp]: If the computer crashes, the file won't get deleted...

**Review by luke-jr — CHANGES_REQUESTED** [review]
I think executing a command is a better approach. You can then have it execute `echo >/path/to/.ready` if you like.

**Comment by practicalswift** [comment]
@murrayn Is it worth writing the PID to the readiness file to allow the consumer to verify that the file is not stale?

Would it be worth to allow for also the "waiting for startup" scenario by writing to say `.bitcoind-state.startup` immediately on startup-up and then rename it to `.bitcoind-state.ready` when the initialization is done?

File: src/init.cpp [hunk 2]
```diff
@@ -120,6 +120,42 @@ NODISCARD static bool CreatePidFile()
...
+    if (file) {
+        fclose(file);
+        return true;
+    } else {
+        return InitError(strprintf(_("Unable to create the ready status file '%s': %s"),
+                   GetReadyFile().string(), std::strerror(errno)));
+    }
+}
+
+static bool RemoveFile(fs::path file)
```
Comment (practicalswift) [inline src/init.cpp]: Should be const ref? :-)

Comment (murrayn) [reply src/init.cpp]: Yes, fixed.

**Comment by murrayn** [comment]
@practicalswift Thanks for the comments. I did consider both of those approaches. But the PID is already available in `bitcoind.pid`, so doing so here seems redundant. I also considered having a **'status'** file of some type, instead of **'ready'**, but I think the only interesting status is likely to be **'ready'** anyway.

File: src/init.cpp (hunk 1 above)
Comment (murrayn) [reply src/init.cpp]: Fair point, but the external command approach would likely have the same limitation. Perhaps we could remove any existing `ready` file early in `AppInitMain()` (or earlier) to mitigate this?

I am very uncomfortable with adding the ability to `exec()` external programs. It just feels like an attack vector.

Comment (luke-jr) [reply src/init.cpp]: `-blocknotify` and `-walletnotify` already exist.

**Comment by promag** [comment]
NACK, sorry.

> the idea is to indicate, via creation of a file, that `bitcoind` is not only running but ready to actually use

I'd say just use it. A good script should handle retrying, not only when it's starting up, but also when it crashed or it's being updated or something. This approach also works if the other service is not in the same host, which this PR can't handle. See the test framework, it checks if it's ready by calling `getblockcount` a couple of times:
https://github.com/bitcoin/bitcoin/blob/2d46f1be0c3c8b7287aa1f62bb1f5b4a8d00ff6e/test/functional/test_framework/test_node.py#L206-L219

**Comment by dongcarl** [comment]
@murrayn I think a better modification is perhaps adding support for `sd_notify` (behind a configure flag) that would tell systemd when things are ready. Or doing something like a systemd socket configuration. But of course all of those are systemd specific.

**Comment by murrayn** [comment]
Doesn't seem to be much traction for this, so I'll close it.

**Pull request closed without merging**